### PR TITLE
 citra_qt: add motion/touch config

### PR DESCRIPTION
**Pull request by zhaowenlan1779**
These functions include reloading udp client, testing communication and configuring calibration. I also added a function to common/thread.h to use WaitFor.

Implemented a Motion/Touch... configuration dialog. Users can open this dialog by clicking a button on the Input tab. The dialog also includes CemuhookUDP server config and communication testing, and the touchpad calibration config.

I had no time to provide a screenshot now, but I will add one soon

UPD: Screenshot
image

Note: Touchpad calibration configuration is not easy to finish according to @jroweboy. I do not have a better idea for this currently though. Probably add a dialog about the process?

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4059)

### CODE REVIEW

File: src/citra_qt/configuration/configure_motion_touch.cpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+            QMetaObject::invokeMethod(this, "ShowCalibrationConfigureResult", Q_ARG(bool, true));
+        },
+        [this] {
+            LOG_ERROR(Frontend, "UDP touchpad calibration config failed");
+            QMetaObject::invokeMethod(this, "ShowCalibrationConfigureResult", Q_ARG(bool, false));
+        });
+}
+
+void ConfigureMotionTouch::closeEvent(QCloseEvent* event) {
+    if (!CanCloseDialog())
```
Comment (lioncash) [inline src/citra_qt/configuration/configure_motion_touch.cpp]: This can be inverted to be in terms of the true case.

File: src/citra_qt/configuration/configure_motion_touch.cpp [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+           "href='https://citra-emu.org/wiki/"
+           "how-to-set-up-your-controller-or-android-phones-for-touch-and-motion-input'><span "
+           "style=\"text-decoration: underline; color:#039be5;\">Learn More</span></a>"));
+
+    setConfiguration();
+    updateUiDisplay();
+    connectEvents();
+}
+
+ConfigureMotionTouch::~ConfigureMotionTouch() {}
```
Comment (lioncash) [inline src/citra_qt/configuration/configure_motion_touch.cpp]:
```cpp

ConfigureMotionTouch::~ConfigureMotionTouch() = default;
```

File: src/input_common/udp/client.cpp [hunk 3]
```diff
@@ -190,4 +192,72 @@ void Client::OnPadData(Response::PadData data) {
...
+        worker_thread.join();
+        if (result)
+            success_callback();
+        else
+            failure_callback();
+    })
+        .detach();
+}
+
+constexpr u16 CALIBRATION_THRESHOLD = 100;
```
Comment (lioncash) [inline src/input_common/udp/client.cpp]: This can be within the function to limit its scope

**Comment by Dartz150** [comment]
Tested, works without any issue until now, only noticed a bit of framedrops in both Majora's Mask and Ocarina of Time while viewing in first person and moving fast the camera, but I remember this was also happening in the original hardware. Is a blast to play citra through Moonlight on Android with this feature enabled, feels like the real deal.

File: src/citra_qt/configuration/configure_motion_touch.cpp [hunk 4]
```diff
@@ -4,13 +4,67 @@
...
+            case CalibrationConfigurationJob::Status::Stage1Completed:
+                text = tr("Now touch the bottom right corner <br>of your touchpad.");
+                break;
+            case CalibrationConfigurationJob::Status::Completed:
+                text = tr("Configuration completed!");
+                break;
+            }
+            QMetaObject::invokeMethod(this, "UpdateLabelText", Q_ARG(QString, text));
+            if (status == CalibrationConfigurationJob::Status::Completed) {
+                cancel_button->setText(tr("OK"));
```
Comment (zhaowenlan1779) [inline src/citra_qt/configuration/configure_motion_touch.cpp]: Note to myself: Only the UI thread can change the UI while this is getting called from a worker thread

Will fix

**Comment by adityaruplaha** [comment]
Note, this supersedes #3929.
